File: Algorithms/leetcode/Ex_121_Best_Time_to_Buy_and_Sell_Stock/Soln_2_One_Pass.java
package Algorithms.leetcode.Ex_121_Best_Time_to_Buy_and_Sell_Stock;

class Solution_2 {
    public int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;   // 记录从0到当前下标所找到的最小价格
        int res = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minPrice) {
                // 随着往后扫，更新最小价格
                minPrice = prices[i];
            } else if (prices[i] - minPrice > res) {
                // 如果当前价格与 prices[0...i-1] 的最小价格之间价差 > 之前的最大利润，则更新res
                res = prices[i] - minPrice;
            }
        }
        return res;
    }
}

// 解法参考：https://leetcode.com/problems/best-time-to-buy-and-sell-stock/solution/

File: Algorithms/leetcode/Ex_0088_Merge_Sorted_Array/Soln_2_One_Pass.cpp
#include <vector>

using std::vector;
using std::swap;

class Solution_2 {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = nums1.size() - 1;
        m = m - 1;
        n = n - 1;
        while (n >= 0) {
            while (m >= 0 && nums1[m] >= nums2[n]) {
                swap(nums1[i], nums1[m]);
                i --;
                m --;
            }
            swap(nums2[n], nums1[i]);
            i --;
            n --;
        }
    }
};

// 参考题解：https://leetcode-cn.com/problems/merge-sorted-array/solution/88-by-ikaruga/
